functions (#8)
Ada-Developers-Academy/core-unit-3: functions/arrow-functions.md

### !end-callout

## Syntax With One-Line Return Function Bodies

Consider our `sayHello` arrow function:

<!-- prettier-ignore-start -->
```javascript
const sayHello = () => {
  return 'hey there.';
}

console.log(sayHello());
```
<!-- prettier-ignore-end -->

This function is responsible for one thing: returning the string `'hey there.'`.

When an arrow function is responsible for returning something, and has a function body that is one line long, we can use reduced arrow function syntax.

<!-- prettier-ignore-start -->
```js
const sayHello = () => 'hey there.';
```
<!-- prettier-ignore-end -->

| <div style="min-width:200px;"> Piece of Code </div> | Notes                                                                                                                                                                       |
| --------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `const sayHello =`                                  | We declare the `const` variable `sayHello` and begin to assign it to an arrow function.                                                                                     |

Reviewer: directionality of assign
```suggestion
| `const sayHello =`                                  | We declare the `const` variable `sayHello` and begin to assign it an arrow function.                                                                                     |
```

| `()`                                                | In one-line arrow functions, the parameter syntax stays the same.                                                                                                           |
| `=>`                                                | In one-line arrow functions, we still need the arrow between the parameter list and the function body.                                                                      |
| `'hey there.'`                                      | In one-line arrow functions, we can exclude the curly braces and the `return` keyword. However, this expression **_must_** be on the same line as the rest of the function. |

> | `;`                                                 | This statement now must conclude with a semicolon.                                                                                                                          |

Reviewer: technically, the previous examples should also have used a semicolon, so this isn't really any different.
```suggestion
| `;`                                                 | This semicolon terminates the expression that assigns the arrow function to `sayHello`.                                                                                                                          |
```

<!-- ### Exercise

Convert this into an arrow function and minimize the code typed.

```javascript
const hello = function () {
  return "hello world";
};
``` -->

## Syntax with Parameters

### With Multiple Parameters or Zero Parameters

Like "normal" functions, arrow functions have their parameters listed in the parentheses.

With zero parameters, we leave the parentheses empty:

```javascript
const sayHelloWorld = () => "Hello World!";
```

With multiple parameters, they are comma-separated.

```javascript
const add = (x, y) => x + y;
```

### With One Parameter

This arrow function takes a single number and returns that number squared:

```javascript
const squareNum = (number) => {
  return number * number;
};
console.log(squareNum(3)); // 9
```

When we have an arrow function that takes in _one and only one_ parameter, the parentheses around the single parameter are optional.

> Consider this refactoring:

Reviewer: ```suggestion
Consider this rewriting:
```

```javascript

> const squareNum = (number) => {

Reviewer: missing the rewrite
```suggestion
const squareNum = number => {
```

  return number * number;
};
console.log(squareNum(3)); // 9
```

> If we don't add parentheses when there are zero or more than one parameter, we'll get a syntax error!

Reviewer: wouldn't hurt to repeat ourselves
```suggestion
Remember that the parentheses are optional _only_ when there is _one and only one_ parameter. If we don't use parentheses when there are zero or more than one parameter, we'll get a syntax error!
```

### !callout-info

> ## Bonus Refactor: One Parameter and One Line
> 
> We can refactor our `squareNum` arrow function, taking advantage of its one parameter and one-line return:

Reviewer: ```suggestion
## Bonus Rewrite: One Parameter and One Line

We can rewrite our `squareNum` arrow function, taking advantage of its one parameter and one-line return:
```

<!-- prettier-ignore-start -->
```js
const squareNum = number => number * number;
```
<!-- prettier-ignore-end -->

### !end-callout

## One-Line Arrow Functions Returning Objects

Consider this `makePerson` function, whose responsibility is to return an object.

```javascript

> const makePerson = function (id, name) {

Reviewer: ```suggestion
const makePerson = function(id, name) {
```

  return {
    id: id,
    name: name,
  };
};
```

> When we refactor it to a one-line arrow function, _because it specifically returns an object_, we need to wrap our object in parentheses.

Reviewer: reword to explain _why_ the braces of an object literal pose a problem.
remove this version an replace with markdown in following comment
no suggestion because code
```suggestion
```

Reviewer: To rewrite it as a one-line arrow function, we might try the following:

```javascript
const makePerson = (id, name) => { id: id, name: name };  // does NOT work!
```

But this will not work as expected!

To JavaScript, the opening brace `{` of the object is indistinguishable from the opening brace we would write at the start of a function body.

Fortunately, we can resolve this ambiguity by wrapping the object literal in parentheses!

```javascript
const makePerson = (id, name) => ({ id: id, name: name });
```

In general, wrapping a value in parentheses is a great way to ensure it's not interpreted as a function body.

### !callout-info

## JavaScript's Object Shorthand

> 

Reviewer: add a more explicit situation
```suggestion

In JavaScript, we'll make many object literals over time. Initializing object literals requires key-value pairs. Notice in the `makePerson` function above that the keys we used in our object happen to match **exactly** the parameter names, resulting in the repetitive `id: id, name: name`.

<br />

```

> In JavaScript, we'll make many object literals over time. Imagine a situation when our object literal needs a key-value pair. When we have **exactly** a _variable_ that references the _value_ we want to include **_and_** that variable is named _exactly_ the name of the key we're inserting, we can use object literal shorthand.

Reviewer: remove the portion that was moved up into the suggested paragraph above
more consistent formatting on the emphasized elements
```suggestion
When we have **exactly** a _variable_ that references the _value_ we want to include **_and_** that variable is named **exactly** the name of the key we're inserting, we can use object literal shorthand.
```

<br/>

> In object literal shorthand, we can comma-separate the variables we want to include in an object. The key-value pairs will form such that the _name_ of the variable is the _key_, and the _value_ of the variable is the _value_.

Reviewer: ```suggestion
In object literal shorthand, we can comma-separate the variables we want to include in an object. The key-value pairs will be formed such that the _name_ of the variable is the _key_, and the _value_ of the variable is the _value_.
```

> <br/>
> 

Reviewer: removal improves the vertical spacing a little
```suggestion
```

<!-- prettier-ignore-start -->
```js
const redFruit = 'apple';
const blueFruit = 'blueberry';

const coolFruitObject = { redFruit, blueFruit };
```
<!-- prettier-ignore-end -->

> 

Reviewer: inclusion improves the vertical spacing
```suggestion

<br/>

```
